cpp/src/arrow/util/decimal.h

```diff
@@ -90,35 +112,85 @@ class ARROW_EXPORT Decimal128 : public BasicDecimal128 {
 
   /// \brief Convert a decimal string to a Decimal128 value, optionally including
   /// precision and scale if they're passed in and not null.
-  static Status FromString(const util::string_view& s, Decimal128* out,
-                           int32_t* precision = NULLPTR, int32_t* scale = NULLPTR);
-  static Status FromString(const std::string& s, Decimal128* out,
-                           int32_t* precision = NULLPTR, int32_t* scale = NULLPTR);
-  static Status FromString(const char* s, Decimal128* out, int32_t* precision = NULLPTR,
-                           int32_t* scale = NULLPTR);
+  static Result<Decimal128> FromString(const util::string_view& s,
+                                       int32_t* precision = NULLPTR,
+                                       int32_t* scale = NULLPTR);
```

Comment:
To avoid ambiguities, perhaps we should have two different API families:
```c++
static Result<Decimal128> FromString(const util::string_view& s);

static Status FromString(const util::string_view& s, Decimal128* out,
                         int32_t* precision, int32_t* scale);

ARROW_DEPRECATED("Use Result-returning version")
static Status FromString(const util::string_view& s, Decimal128* out);
```